File: services/gnss-ingestor/internal/ntrip/handshake.go
package ntrip

import (
	"bytes"
	"fmt"
	"strconv"
	"strings"
)

// NTRIP v2 is spoken first; v1-compatible ICY 200 responses are tolerated.
// Documented accepted successes: "HTTP/1.x 200" and "ICY 200".

// FailureClass classifies handshake/stream failures for fail-closed handling.
type FailureClass string

const (
	FailureAuth         FailureClass = "AUTH_FAILURE"
	FailureForbidden    FailureClass = "AUTHORIZATION_REJECTED"
	FailureNotFound     FailureClass = "MOUNTPOINT_NOT_FOUND"
	FailureConfig       FailureClass = "CONFIG_INVALID"
	FailureProtocol     FailureClass = "PROTOCOL_ERROR"
	FailureWrongContent FailureClass = "WRONG_CONTENT"
	FailureEmpty        FailureClass = "EMPTY_RESPONSE"
	FailureTimeout      FailureClass = "TIMEOUT"
	FailureStalled      FailureClass = "STALLED"
	FailureLost         FailureClass = "CONNECTION_LOST"
	FailureNMEARequired FailureClass = "NMEA_POSITION_REQUIRED"
)

// TerminalFailures must never reconnect without a configuration change:
// bad credentials, unknown mountpoint, explicit rejection, bad config.
func IsTerminal(failure FailureClass) bool {
	switch failure {
	case FailureAuth, FailureNotFound, FailureForbidden, FailureConfig:
		return true
	default:
		return false
	}
}

// HandshakeResult is a validated (or refused) stream handshake.
type HandshakeResult struct {
	OK         bool
	Failure    FailureClass
	Reason     string
	StatusCode int
}

// BuildStreamRequest renders a GET request for the mountpoint (NTRIP v2 first).
func BuildStreamRequest(cfg Config) []byte {
	mount := cfg.Mountpoint
	if !strings.HasPrefix(mount, "/") {
		mount = "/" + mount
	}
	var sb strings.Builder
	fmt.Fprintf(&sb, "GET %s HTTP/1.1\r\n", mount)
	fmt.Fprintf(&sb, "Host: %s:%d\r\n", cfg.Host, cfg.Port)
	sb.WriteString("Ntrip-Version: Ntrip/2.0\r\n")
	fmt.Fprintf(&sb, "User-Agent: %s\r\n", cfg.UserAgent)
	sb.WriteString("Accept: */*\r\nConnection: close\r\n")
	if cfg.Username != "" {
		fmt.Fprintf(&sb, "Authorization: %s\r\n", BasicAuthHeader(cfg.Username, cfg.Password))
	}
	if strings.TrimSpace(cfg.GGASentence) != "" {
		fmt.Fprintf(&sb, "Ntrip-GGA: %s\r\n", strings.TrimSpace(cfg.GGASentence))
	}
	sb.WriteString("\r\n")
	return []byte(sb.String())
}

// BuildSourcetableRequest renders a source-table request (GET /).
func BuildSourcetableRequest(cfg Config) []byte {
	var sb strings.Builder
	sb.WriteString("GET / HTTP/1.1\r\n")
	fmt.Fprintf(&sb, "Host: %s:%d\r\n", cfg.Host, cfg.Port)
	sb.WriteString("Ntrip-Version: Ntrip/2.0\r\n")
	fmt.Fprintf(&sb, "User-Agent: %s\r\n", cfg.UserAgent)
	sb.WriteString("Accept: */*\r\nConnection: close\r\n")
	if cfg.Username != "" {
		fmt.Fprintf(&sb, "Authorization: %s\r\n", BasicAuthHeader(cfg.Username, cfg.Password))
	}
	sb.WriteString("\r\n")
	return []byte(sb.String())
}

// ParseStatusLine splits a status line into protocol, code (0 if absent), text.
func ParseStatusLine(line string) (string, int, string) {
	parts := strings.SplitN(strings.TrimSpace(line), " ", 3)
	if len(parts) < 2 {
		return "", 0, strings.TrimSpace(line)
	}
	code, err := strconv.Atoi(parts[1])
	if err != nil {
		return parts[0], 0, strings.TrimSpace(line)
	}
	text := ""
	if len(parts) > 2 {
		text = parts[2]
	}
	return parts[0], code, text
}

func looksLikeSourcetable(prefix []byte) bool {
	if len(prefix) > 512 {
		prefix = prefix[:512]
	}
	text := string(prefix)
	upper := strings.ToUpper(text)
	return strings.HasPrefix(text, "STR;") || strings.HasPrefix(text, "CAS;") ||
		strings.HasPrefix(text, "NET;") || strings.Contains(upper, "ENDSOURCETABLE")
}

func looksLikeHTML(prefix []byte) bool {
	if len(prefix) > 1024 {
		prefix = prefix[:1024]
	}
	lowered := strings.ToLower(strings.TrimSpace(string(prefix)))
	return strings.HasPrefix(lowered, "<html") || strings.HasPrefix(lowered, "<!doctype")
}

// ValidateStreamResponse validates a stream handshake, failing closed so
// no binary parser ever consumes an HTTP error body as RTCM.
func ValidateStreamResponse(statusLine string, headers map[string]string, bodyPrefix []byte) HandshakeResult {
	_ = headers
	if strings.TrimSpace(statusLine) == "" {
		return HandshakeResult{Failure: FailureEmpty, Reason: "empty status line"}
	}
	protocol, code, _ := ParseStatusLine(statusLine)
	known := protocol == "HTTP/1.0" || protocol == "HTTP/1.1" || protocol == "ICY" ||
		strings.HasPrefix(strings.ToUpper(strings.TrimSpace(statusLine)), "ICY ")
	if !known || code == 0 {
		return HandshakeResult{Failure: FailureProtocol, Reason: "malformed status line: " + statusLine}
	}
	switch code {
	case 200:
		if looksLikeHTML(bodyPrefix) {
			return HandshakeResult{Failure: FailureWrongContent, Reason: "HTML body where RTCM expected", StatusCode: code}
		}
		if looksLikeSourcetable(bodyPrefix) {
			return HandshakeResult{Failure: FailureWrongContent, Reason: "source-table body where RTCM stream expected", StatusCode: code}
		}
		return HandshakeResult{OK: true, Reason: "handshake_success", StatusCode: code}
	case 401:
		return HandshakeResult{Failure: FailureAuth, Reason: "authorization rejected: 401", StatusCode: code}
	case 403:
		return HandshakeResult{Failure: FailureForbidden, Reason: "authorization rejected: 403", StatusCode: code}
	case 404:
		return HandshakeResult{Failure: FailureNotFound, Reason: "mountpoint not found: 404", StatusCode: code}
	default:
		return HandshakeResult{Failure: FailureProtocol, Reason: "unexpected status: " + strconv.Itoa(code), StatusCode: code}
	}
}

// SplitHead splits raw bytes into head and rest at the first blank line.
// It returns ok=false when no terminator is present yet (or on oversize).
func SplitHead(data []byte, maxHeaderBytes int) (head, rest []byte, ok bool) {
	for _, sep := range [][]byte{[]byte("\r\n\r\n"), []byte("\n\n")} {
		if index := bytes.Index(data, sep); index >= 0 {
			head = data[:index]
			if len(head) > maxHeaderBytes {
				return nil, nil, false
			}
			return head, data[index+len(sep):], true
		}
	}
	if len(data) > maxHeaderBytes {
		return nil, nil, false
	}
	return nil, nil, false
}

// ParseHead parses a response head into status line and headers.
func ParseHead(head []byte) (string, map[string]string) {
	lines := strings.Split(string(head), "\n")
	status := ""
	if len(lines) > 0 {
		status = strings.TrimRight(lines[0], "\r")
	}
	headers := map[string]string{}
	for _, line := range lines[1:] {
		line = strings.TrimRight(line, "\r")
		if key, value, found := strings.Cut(line, ":"); found {
			headers[strings.TrimSpace(key)] = strings.TrimSpace(value)
		}
	}
	return status, headers
}

File: research/ntrip_ingestion/src/nlgcp_ntrip_ingest/handshake.py
"""NTRIP v2-first HTTP handshake with v1 tolerance.

Accepted success responses: ``HTTP/1.x 200`` and ``ICY 200`` (v1
casters).  Everything else is classified fail-closed; binary parsers
must never consume HTTP error bodies as RTCM.
"""

from __future__ import annotations

from dataclasses import dataclass

from nlgcp_ntrip_ingest.auth import basic_auth_header
from nlgcp_ntrip_ingest.models import FailureClass, NtripConfig

NTRIP_V2 = "Ntrip-Version: Ntrip/2.0"


@dataclass(frozen=True, slots=True)
class HandshakeResult:
    ok: bool
    failure: FailureClass | None
    reason: str
    status_code: int | None


def build_stream_request(config: NtripConfig) -> bytes:
    """Build a GET request for the configured mountpoint (NTRIP v2 first)."""
    mount = config.mountpoint if config.mountpoint.startswith("/") else "/" + config.mountpoint
    lines = [
        f"GET {mount} HTTP/1.1",
        f"Host: {config.host}:{config.port}",
        NTRIP_V2,
        f"User-Agent: {config.user_agent}",
        "Accept: */*",
        "Connection: close",
    ]
    if config.username:
        lines.append(f"Authorization: {basic_auth_header(config.username, config.password)}")
    if config.gga_sentence:
        lines.append(f"Ntrip-GGA: {config.gga_sentence.strip()}")
    return ("\r\n".join(lines) + "\r\n\r\n").encode("ascii", errors="replace")


def build_sourcetable_request(config: NtripConfig) -> bytes:
    """Build a source-table request (``GET /`` with Ntrip-Version)."""
    lines = [
        "GET / HTTP/1.1",
        f"Host: {config.host}:{config.port}",
        NTRIP_V2,
        f"User-Agent: {config.user_agent}",
        "Accept: */*",
        "Connection: close",
    ]
    if config.username:
        lines.append(f"Authorization: {basic_auth_header(config.username, config.password)}")
    return ("\r\n".join(lines) + "\r\n\r\n").encode("ascii", errors="replace")


def parse_status_line(line: str) -> tuple[str, int | None, str]:
    """Split a status line into (protocol, code, text); code may be None."""
    parts = line.strip().split(None, 2)
    if len(parts) < 2:
        return ("", None, line.strip())
    protocol = parts[0]
    try:
        return (protocol, int(parts[1]), parts[2] if len(parts) > 2 else "")
    except ValueError:
        return (protocol, None, line.strip())


def _looks_like_sourcetable(body_prefix: bytes) -> bool:
    text = body_prefix[:512].decode("latin-1", errors="replace")
    upper = text.upper()
    return (
        text.startswith("STR;")
        or "ENDSOURCETABLE" in upper
        or text.startswith("CAS;")
        or text.startswith("NET;")
    )


def _looks_like_html(body_prefix: bytes) -> bool:
    lowered = body_prefix[:1024].decode("latin-1", errors="replace").strip().lower()
    return lowered.startswith("<html") or lowered.startswith("<!doctype")


def validate_stream_response(
    status_line: str, headers: dict[str, str], body_prefix: bytes
) -> HandshakeResult:
    """Validate a stream handshake response, failing closed on misuse."""
    if not status_line.strip():
        return HandshakeResult(False, FailureClass.EMPTY_RESPONSE, "empty status line", None)
    protocol, code, _ = parse_status_line(status_line)
    normalised = status_line.strip().upper()
    known = protocol.upper() in ("HTTP/1.0", "HTTP/1.1", "ICY")
    if not known and not normalised.startswith("ICY "):
        return HandshakeResult(
            False, FailureClass.PROTOCOL_ERROR, f"malformed status line: {status_line!r}", code
        )
    if code is None:
        return HandshakeResult(
            False, FailureClass.PROTOCOL_ERROR, f"malformed status line: {status_line!r}", None
        )
    if code == 200:
        if _looks_like_html(body_prefix):
            return HandshakeResult(
                False, FailureClass.WRONG_CONTENT, "HTML body where RTCM expected", code
            )
        if _looks_like_sourcetable(body_prefix):
            return HandshakeResult(
                False,
                FailureClass.WRONG_CONTENT,
                "source-table body where RTCM stream expected",
                code,
            )
        return HandshakeResult(True, None, "handshake_success", code)
    if code in (401, 403):
        failure = (
            FailureClass.AUTH_FAILURE if code == 401 else FailureClass.AUTHORIZATION_REJECTED
        )
        return HandshakeResult(False, failure, f"authorization rejected: {code}", code)
    if code == 404:
        return HandshakeResult(
            False, FailureClass.MOUNTPOINT_NOT_FOUND, "mountpoint not found: 404", code
        )
    return HandshakeResult(
        False, FailureClass.PROTOCOL_ERROR, f"unexpected status: {code}", code
    )


def split_head(data: bytes, max_header_bytes: int = 8192) -> tuple[bytes, bytes] | None:
    """Split raw bytes into (head, rest) at the first blank line."""
    for sep in (b"\r\n\r\n", b"\n\n"):
        index = data.find(sep)
        if index >= 0:
            head = data[:index]
            if len(head) > max_header_bytes:
                return None
            return (head, data[index + len(sep):])
    if len(data) > max_header_bytes:
        return None
    return None


def parse_head(head: bytes) -> tuple[str, dict[str, str]]:
    """Parse a response head into (status_line, headers)."""
    text = head.decode("latin-1", errors="replace")
    lines = text.splitlines()
    status = lines[0] if lines else ""
    headers: dict[str, str] = {}
    for line in lines[1:]:
        if ":" in line:
            key, value = line.split(":", 1)
            headers[key.strip()] = value.strip()
    return status, headers
